cpp/include/cudf/dictionary/dictionary_column_view.hpp

```diff
@@ -0,0 +1,228 @@
+                                            rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/**
+ * @brief Create a new dictionary column by applying only the specified keys
+ * to the existing dictionary_column.
+ *
+ * Any new elements found in the keys parameter are added to the output dictionary.
+ * Any existing keys not in the keys parameter are removed.
+ *
+ * The indices are updated to reflect the position values of the new keys.
+ * Any indices pointing to removed keys are set to null.
+ *
+ * ```
+ * d1 = {["a","c","d"],[2,0,1,0]}
+ * d2 = set_keys(d1,["a","b","c"])
+ * d2 is now {["a","b","c"],[null,0,2,0]}
+ * ```
+ *
+ * @throw cudf_logic_error if the keys type does not match the keys type in
+ * the dictionary_column.
+ *
+ * @param dictionary_column Existing dictionary column.
+ * @param keys New keys to use for the output column.
+ * @param mr Resource for allocating memory for the output.
+ * @return New dictionary column.
+ */
+std::unique_ptr<column> set_keys( dictionary_column_view const& dictionary_column,
+                                  column_view const& keys,
+                                  rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/**
+ * @brief Create a new dictionary column by merging the keys and indices
+ * from two existing dictionary columns.
+ *
+ * The indices of the resulting column are created to appear as though the
+ * second dictionary_column indices is appended to the first.
+ *
+ * ```
+ * d1 = {["a","c","d"],[2,0,1,0]}
+ * d2 = {["b","e"],[1,0,1,1,0]}
+ * d3 = merge(d1,d2)
+ * d3 is now {["a","b","c","d","e"],[3,0,2,0,4,1,4,4,1]}
+ * ```
+ *
+ * @throw cudf_logic_error if the keys types do not match for both dictionary columns.
+ *
+ * @param dictionary_column1 Existing dictionary column.
+ * @param dictionary_column2 2nd existing dictionary column.
+ * @param mr Resource for allocating memory for the output.
+ * @return New dictionary column.
+ */
+std::unique_ptr<column> merge_dictionaries( dictionary_column_view const& dictionary_column1,
+                                            dictionary_column_view const& dictionary_column2,
+                                            rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/**
+ * @brief Create a column by gathering the keys from the provided
+ * dictionary_column into a new column using the specified indices.
+ *
+ * ```
+ * d1 = {["a","c","d"],[2,0,1,0]}
+ * s = gather_type(d1,[1,0,0,2,2])
+ * s is now ["c","a","a","d","d"]
+ * ```
+ *
+ * @param dictionary_column Existing dictionary column.
+ * @param indices Indices to use for the new dictionary column.
+ * @param mr Resource for allocating memory for the output.
+ * @return New column with type matching the dictionary_column's keys.
+ */
+std::unique_ptr<column> gather_type( dictionary_column_view const& dictionary_column,
```

Comment:
Just that the intermediate dictionary-column does not need to be created.
I was thinking of having `to_type` just call this one passing all of its indices intact.
Or something like that.

Reply:
Did you have a usecase in mind for `gather_type`?

Reply:
> Did you have a usecase in mind for `gather_type`?

Well, I had to create 3 `gather` methods for `NVCategory`.  Two of them create new dictionary columns. The _normal_ gather will be handled by regular `cudf::gather`. The 2nd `gather` will be handled by a factory method. 

The third `gather` returns a column of the keys' type and not a full dictionary. 
https://github.com/rapidsai/cudf/blob/6a5a801139a9a379cc42435e31806dff10e0b209/cpp/include/nvstrings/NVCategory.h#L334

I'm assuming since these 3 `gather` methods were created for some need in `NVCategory` and so they will be needed in the ported dictionary column as well.

Reply:
Let's strive not to add anything until there is an immediate need for it to prevent adding things we may not actually need.